zcash_client_sqlite/src/testing/pool.rs

```diff
@@ -1599,6 +1613,212 @@ pub(crate) fn fully_funded_fully_private<P0: ShieldedPoolTester, P1: ShieldedPoo
+        st.generate_empty_block();
+    }
+
+    // Scan into the middle of the empty range
+    let scanned = st.scan_cached_blocks(next_to_scan, 5);
+    let next_to_scan = scanned.scanned_range().end;
+
+    let mid_balance = (note_value * 3).unwrap();
+    assert_eq!(st.get_total_balance(acct_id), mid_balance);
+    assert_eq!(st.get_spendable_balance(acct_id, 1), mid_balance);
+
+    // Set up the fee rule and input selector we'll use for all the transfers.
+    let fee_rule = StandardFeeRule::Zip317;
+    let input_selector = GreedyInputSelector::new(
+        standard::SingleOutputChangeStrategy::new(fee_rule, None, P1::SHIELDED_PROTOCOL),
+        DustOutputPolicy::default(),
+    );
+
+    // First, send funds just to P0
+    let transfer_amount = NonNegativeAmount::const_from_u64(200000);
+    let p0_transfer = zip321::TransactionRequest::new(vec![Payment::without_memo(
+        P0::random_address(&mut st.rng),
+        transfer_amount,
+    )])
+    .unwrap();
+    let res = st
+        .spend(
+            &input_selector,
+            account.usk(),
+            p0_transfer,
+            OvkPolicy::Sender,
+            NonZeroU32::new(1).unwrap(),
+        )
+        .unwrap();
+    st.generate_next_block_including(*res.first());
+
+    let expected_fee = NonNegativeAmount::const_from_u64(10000);
+    let expected_change = (note_value - transfer_amount - expected_fee).unwrap();
+    assert_eq!(
+        st.get_total_balance(acct_id),
+        (initial_balance + expected_change).unwrap()
+    );
+    assert_eq!(st.get_pending_change(acct_id, 1), expected_change);
+
+    // In the next block, send funds to both P0 and P1
+    let both_transfer = zip321::TransactionRequest::new(vec![
+        Payment::without_memo(P0::random_address(&mut st.rng), transfer_amount),
+        Payment::without_memo(P1::random_address(&mut st.rng), transfer_amount),
+    ])
+    .unwrap();
+    let res = st
+        .spend(
+            &input_selector,
+            account.usk(),
+            both_transfer,
+            OvkPolicy::Sender,
+            NonZeroU32::new(1).unwrap(),
+        )
+        .unwrap();
+    st.generate_next_block_including(*res.first());
+
+    // Generate a few more empty blocks
+    for _ in 0..5 {
+        st.generate_empty_block();
+    }
+
+    // Generate another block with funds for us
+    let (max_height, _, _) =
+        st.generate_next_block(&p0_fvk, AddressType::DefaultExternal, note_value);
+
+    // Scan everything.
+    st.scan_cached_blocks(
+        next_to_scan,
+        usize::try_from(u32::from(max_height) - u32::from(next_to_scan) + 1).unwrap(),
+    );
+
+    let expected_final =
+        (mid_balance + note_value - (transfer_amount * 3).unwrap() - (expected_fee * 3).unwrap())
+            .unwrap();
+    assert_eq!(st.get_total_balance(acct_id), expected_final);
+
+    use incrementalmerkletree::Position;
+    let expected_checkpoints_p0: Vec<(BlockHeight, ShieldedProtocol, Option<Position>)> = [
```

Comment:
It would be useful to add some inline comments to the test above that describe how the tree state is being manipulated by each choice of scan range, which lead to these expected checkpoints.